src/Build/Instance/HostServices.cs

```diff
@@ -69,9 +76,30 @@ public ITaskHost GetHostObject(string projectFile, string targetName, string tas
                 return null;
             }
 
-            ITaskHost hostObject = hostObjects.GetAnyMatchingHostObject(targetName, taskName);
+            var monikerNameOrITaskHost =
+                hostObjects.GetAnyMatchingMonikerNameOrITaskHost(targetName, taskName);
 
-            return hostObject;
+            if (monikerNameOrITaskHost == null)
+            {
+                return null;
+            }
+            else
+            {
+                if (monikerNameOrITaskHost.IsMoniker)
+                {
+#if FEATURE_COM_INTEROP
+                    object objectFromRunningObjectTable =
+                        _runningObjectTable.GetObject(monikerNameOrITaskHost.MonikerName);
+                    return (ITaskHost)objectFromRunningObjectTable;
```

Comment:
Are COM exceptions here going to bubble up nicely?

Thinking about all of the failures I had while developing this, there wasn't a heap of information in the build log (ie, what moniker, what task, what target) though that could have been the Project System Tools UI not handling something.

Reply:
Yeah, thinking through and at least manually testing failure cases here is a great idea. Ideally we'd get checked-in tests but I don't know how hard it is to get the COM objects to do the testing with/set up our test infra to do it.

Reply:
> bubble up nicely

I could try to wrap exceptions that may happen from this 2 operations and add moniker, task, target. What other info do you need? Although I think this error should be seen by us only unless the user broke the windows registry. 

> checked-in tests

That is one of the annoying problem. The most fragile part -- CPS, msbuild and the task (from xaml) have all interfaces aligned and they are registered. This need a whole VS setup from clean machine to test.

Since I mocked out runningObjectTable, serialization/deserialization and "put it in msbuild internal dictionary" are covered. Only thing that is not covered is RunningObjectTable.cs and it is very shallow.

If we add end to end test, the best we can have is extra coverage on RunningObjectTable.cs. But we still leave the interface align part off.

Reply:
> I could try to wrap exceptions that may happen from this 2 operations and add moniker, task, target. What other info do you need?

That would be enough to investigate I'm sure, especially if we get failures from custom targets.

The other question, does an error here fail the whole build, or just the task? or target?

Reply:
@rainersigwald correct me if i am wrong. There is nothing to catch this exception, so it will fail the whole build. And that is the same experience during testing

Reply:
This gets called from inside a task calling `buildEngine.GetHostObject()` right? In that case, an uncaught exception bubbles up to the task-running code, which catches it, writes an error, and fails the task. The details of what happen next can vary (you can run any task with `ContinueOn="ErrorAndContinue"` or `WarnAndContinue`) but generally a failed task stops execution of the target and marks the target as failed, which stops execution of the project.